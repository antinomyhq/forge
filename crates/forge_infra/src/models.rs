use chrono::NaiveDateTime;
use diesel::prelude::*;
use serde::{Deserialize, Serialize};

use crate::schema::conversations;

#[derive(Debug, Queryable, Selectable, Serialize, Deserialize, Clone)]
#[diesel(table_name = conversations)]
#[diesel(check_for_backend(diesel::sqlite::Sqlite))]
pub struct ConversationRecord {
    pub conversation_id: String,
    pub title: Option<String>,
    pub workspace_id: String,
    pub context: String,
    pub created_at: NaiveDateTime,
    pub updated_at: Option<NaiveDateTime>,
}

#[derive(Debug, Insertable, Serialize, Deserialize)]
#[diesel(table_name = conversations)]
pub struct NewConversationRecord {
    pub conversation_id: String,
    pub title: Option<String>,
    pub workspace_id: String,
    pub context: String,
}

#[derive(Debug, Insertable, Serialize, Deserialize)]
#[diesel(table_name = conversations)]
pub struct UpsertConversationRecord {
    pub conversation_id: String,
    pub title: Option<String>,
    pub workspace_id: String,
    pub context: String,
    pub updated_at: Option<NaiveDateTime>,
}

impl TryFrom<&forge_domain::Conversation> for NewConversationRecord {
    type Error = anyhow::Error;

    fn try_from(conversation: &forge_domain::Conversation) -> Result<Self, Self::Error> {
        let context = serde_json::to_string(conversation)?;
        let workspace_id = &conversation.workspace_id;
        Ok(NewConversationRecord {
            title: conversation.title.clone(),
            conversation_id: conversation.id.into_string(),
            workspace_id: workspace_id.to_string(),
            context,
        })
    }
}

impl TryFrom<&forge_domain::Conversation> for UpsertConversationRecord {
    type Error = anyhow::Error;

    fn try_from(conversation: &forge_domain::Conversation) -> Result<Self, Self::Error> {
        let context = serde_json::to_string(&conversation)?;
        let workspace_id = &conversation.workspace_id;

        Ok(UpsertConversationRecord {
            title: conversation.title.clone(),
            conversation_id: conversation.id.into_string(),
            workspace_id: workspace_id.to_string(),
            context,
            updated_at: Some(chrono::Utc::now().naive_utc()),
        })
    }
}

impl TryFrom<&ConversationRecord> for forge_domain::Conversation {
    type Error = anyhow::Error;

    fn try_from(record: &ConversationRecord) -> Result<Self, Self::Error> {
        let conversation: forge_domain::Conversation = serde_json::from_str(&record.context)?;
        Ok(conversation)
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashMap;

    use chrono::Utc;
    use forge_domain::{Agent, AgentId, ConversationId, Event, Workflow, WorkspaceId};
    use pretty_assertions::assert_eq;
    use serde_json::Value;

    use super::*;

    #[test]
    fn test_try_from_domain_conversation() {
        let fixture = create_test_conversation();
        let actual =
            NewConversationRecord::try_from(&fixture).unwrap();

        assert_eq!(actual.conversation_id, fixture.id.into_string());
        assert_eq!(actual.workspace_id, fixture.workspace_id.to_string());
        assert!(!actual.context.is_empty());

        // Verify we can deserialize back to conversation
        let deserialized: forge_domain::Conversation =
            serde_json::from_str(&actual.context).unwrap();
        assert_eq!(deserialized.id, fixture.id);
    }
    #[test]
    fn test_try_from_domain_conversation_upsert() {
        let fixture = create_test_conversation();
        let actual =
            UpsertConversationRecord::try_from(&fixture).unwrap();

        assert_eq!(actual.conversation_id, fixture.id.into_string());
        assert_eq!(actual.workspace_id, fixture.workspace_id.to_string());
        assert!(!actual.context.is_empty());

        // Verify we can deserialize back to conversation
        let deserialized: forge_domain::Conversation =
            serde_json::from_str(&actual.context).unwrap();
        assert_eq!(deserialized.id, fixture.id);

        // Verify updated_at is set to recent time (within last second)
        let now = Utc::now().naive_utc();
        let updated_at = actual.updated_at.expect("updated_at should be set in upsert operation");
        let time_diff = (now - updated_at).num_seconds();
        assert!(time_diff >= 0 && time_diff <= 1);
    }
    #[test]
    fn test_conversation_repository_init() {
        use tempfile::TempDir;

        let temp_dir = TempDir::new().unwrap();
        let db_path = temp_dir.path().join("test.db");

        // Should create and initialize in one step
        let repo = super::super::ConversationRepository::init(db_path.clone()).unwrap();

        // Database file should exist
        assert!(db_path.exists());

        // Should be able to get a connection after initialization
        let connection_result = repo.get_connection();
        assert!(connection_result.is_ok());
    }

    #[test]
    fn test_try_from_conversation_record() {
        let fixture = create_test_conversation();
        let record =
            NewConversationRecord::try_from(&fixture).unwrap();
        let record = ConversationRecord {
            conversation_id: record.conversation_id,
            title: record.title,
            workspace_id: record.workspace_id,
            context: record.context,
            created_at: Utc::now().naive_utc(),
            updated_at: None,
        };

        let actual = forge_domain::Conversation::try_from(&record).unwrap();
        assert_eq!(actual.id, fixture.id);
        assert_eq!(actual.agents.len(), fixture.agents.len());
        assert_eq!(actual.events.len(), fixture.events.len());
    }

    fn create_test_conversation() -> forge_domain::Conversation {
        let id = ConversationId::generate();
        let mut conversation = forge_domain::Conversation::new(
            id,
            WorkspaceId::new("a1b2c3d4e5f6789a"),
            Workflow::new(),
            vec![],
            vec![Agent::new(AgentId::default())],
        );

        let mut event_data = HashMap::new();
        event_data.insert(
            "content".to_string(),
            Value::String("Hello, world!".to_string()),
        );

        conversation.insert_event(Event {
            id: "test-event-1".to_string(),
            name: "user_task".to_string(),
            value: Some(Value::Object(event_data.into_iter().collect())),
            timestamp: chrono::Utc::now().to_rfc3339(),
            attachments: vec![],
        });

        conversation
    }
}
