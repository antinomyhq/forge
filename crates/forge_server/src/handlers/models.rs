use std::sync::Arc;

use axum::extract::State;
use axum::Json;
use forge_api::API;
use forge_domain::{Model, Services};
use forge_services::Infrastructure;

use super::app_state::AppState;
use crate::error::Result;

/// Handler for models
pub async fn models<F: Services + Infrastructure>(
    State(state): State<Arc<AppState<F>>>,
) -> Result<Json<Vec<Model>>> {
    let models = state.api.models().await?;
    Ok(Json(models))
}
