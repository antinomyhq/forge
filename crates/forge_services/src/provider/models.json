[
  {
    "id": "forge",
    "api_key_vars": "FORGE_API_KEY",
    "url_param_vars": [],
    "response_type": "OpenAI",
    "url": "https://antinomy.ai/api/v1/"
  },
  {
    "id": "open_router",
    "api_key_vars": "OPENROUTER_API_KEY",
    "url_param_vars": [],
    "response_type": "OpenAI",
    "url": "https://openrouter.ai/api/v1/"
  },
  {
    "id": "requesty",
    "api_key_vars": "REQUESTY_API_KEY",
    "url_param_vars": [],
    "response_type": "OpenAI",
    "url": "https://router.requesty.ai/v1/"
  },
  {
    "id": "xai",
    "api_key_vars": "XAI_API_KEY",
    "url_param_vars": [],
    "response_type": "OpenAI",
    "url": "https://api.x.ai/v1/"
  },
  {
    "id": "openai",
    "api_key_vars": "OPENAI_API_KEY",
    "url_param_vars": [],
    "response_type": "OpenAI",
    "url": "https://api.openai.com/v1/"
  },
  {
    "id": "anthropic",
    "api_key_vars": "ANTHROPIC_API_KEY",
    "url_param_vars": [],
    "response_type": "Anthropic",
    "url": "https://api.anthropic.com/v1/"
  },
  {
    "id": "cerebras",
    "api_key_vars": "CEREBRAS_API_KEY",
    "url_param_vars": [],
    "response_type": "OpenAI",
    "url": "https://api.cerebras.ai/v1/"
  },
  {
    "id": "zai",
    "api_key_vars": "ZAI_API_KEY",
    "url_param_vars": [],
    "response_type": "OpenAI",
    "url": "https://api.z.ai/api/paas/v4/"
  },
  {
    "id": "zai_coding",
    "api_key_vars": "ZAI_CODING_API_KEY",
    "url_param_vars": [],
    "response_type": "OpenAI",
    "url": "https://api.z.ai/api/coding/paas/v4/"
  },
  {
    "id": "vertex_ai",
    "api_key_vars": "VERTEX_AI_AUTH_TOKEN",
    "url_param_vars": ["PROJECT_ID", "LOCATION"],
    "response_type": "OpenAI",
    "url": "{{#if (eq location \"global\")}}https://aiplatform.googleapis.com/v1/projects/{{project_id}}/locations/{{location}}/endpoints/openapi/{{else}}https://{{location}}-aiplatform.googleapis.com/v1/projects/{{project_id}}/locations/{{location}}/endpoints/openapi/{{/if}}"
  }
]