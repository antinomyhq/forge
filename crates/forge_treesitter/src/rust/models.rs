use std::path::PathBuf;

/// Represents the type of code block captured
#[derive(Debug, Clone, PartialEq)]
pub enum Kind {
    Function,
    Method,
    Struct,
    Enum,
    Impl,
    Trait,
    Constant,
    //... and so on
}

/// Represents a specific scope for the code block
#[derive(Debug, Clone, PartialEq)]
pub enum Scope {
    Test,
}

/// Represents a captured code block with metadata
#[derive(Debug, Clone, PartialEq)]
pub struct Block {
    /// what type of snippet captured
    kind: Kind,
    /// code file path
    path: PathBuf,
    /// code snippet
    snippet: String,
    /// scope of kind
    scope: Option<Scope>,
    /// Span of snippet
    span: Span,
}

impl Block {
    /// Create a new code block
    pub fn new(
        kind: Kind,
        path: PathBuf,
        snippet: String,
        scope: Option<Scope>,
        span: Span,
    ) -> Self {
        Self {
            kind,
            path,
            snippet,
            scope,
            span,
        }
    }
}

/// Represents a source code span with start and end locations
#[derive(Debug, Clone, PartialEq)]
pub struct Span {
    pub start: Location,
    pub end: Location,
}

/// Represents a location in source code with line and column
#[derive(Debug, Clone, PartialEq)]
pub struct Location {
    pub line: usize,
    pub column: usize,
}

impl From<tree_sitter::Point> for Location {
    fn from(point: tree_sitter::Point) -> Self {
        Self { line: point.row, column: point.column }
    }
}