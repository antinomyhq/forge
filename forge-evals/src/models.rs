// Defines the core data structures for our conversations and tests.
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Clone)]
pub enum Role {
    User,
    Assistant,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Message {
    pub role: Role,
    pub content: String,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct TestCase {
    pub id: String,
    pub conversation: Vec<Message>,
    // The question to ask the agent to test information retrieval post-compaction.
    pub retrieval_test_question: String,
    // The expected answer or a keyword the answer should contain.
    pub expected_answer_keyword: String,
}
