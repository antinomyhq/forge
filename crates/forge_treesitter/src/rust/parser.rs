use std::{fmt::Display, path::Path};

use super::models::{Block, Kind, Location, Offset, Span};
use tree_sitter::StreamingIterator;

const QUERIES: &'static str = include_str!("../queries/rust.scm");

/// Represents a query pattern with its associated metadata
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Pattern {
    Struct,
    Enum,
    Function,
}

impl Display for Pattern {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let str_repr = match self {
            Pattern::Struct => "struct.definition",
            Pattern::Enum => "enum.definition",
            Pattern::Function => "function.definition",
        };
        write!(f, "{}", str_repr)
    }
}

impl TryFrom<&str> for Pattern {
    type Error = anyhow::Error;
    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value {
            "struct.definition" => Ok(Pattern::Struct),
            "enum.definition" => Ok(Pattern::Enum),
            "function.definition" => Ok(Pattern::Function),
            _ => Err(anyhow::anyhow!("Unsupported pattern found '{}'", value)),
        }
    }
}

pub struct RustTreeSitter(tree_sitter::Parser);

impl Default for RustTreeSitter {
    fn default() -> Self {
        let mut parser = tree_sitter::Parser::new();
        parser
            .set_language(&tree_sitter_rust::LANGUAGE.into())
            .expect("shouldn't fail.");

        Self(parser)
    }
}

impl From<Pattern> for Kind {
    fn from(value: Pattern) -> Self {
        match value {
            Pattern::Struct => Kind::Struct,
            Pattern::Enum => Kind::Enum,
            Pattern::Function => Kind::Function,
        }
    }
}

impl RustTreeSitter {
    pub fn parse(&mut self, path: impl AsRef<Path>, source: &str) -> anyhow::Result<Vec<Block>> {
        let tree = self
            .0
            .parse(source, None)
            .ok_or(anyhow::anyhow!("Failed to parse code"))?;

        let query = tree_sitter::Query::new(&tree_sitter_rust::LANGUAGE.into(), QUERIES)?;

        // Execute query
        let mut cursor = tree_sitter::QueryCursor::new();
        let mut matches = cursor.matches(&query, tree.root_node(), source.as_bytes());

        let path_buf = path.as_ref().to_path_buf();
        let mut blocks = Vec::new();

        while let Some(match_) = matches.next() {
            // Get the matched node and its kind
            let capture = match_.captures[0];
            let node = capture.node;

            // Get the name of the capture from the query
            let capture_name = query.capture_names()[capture.index as usize];

            // Convert the capture name to a Pattern and then to a Kind
            let pattern = match Pattern::try_from(capture_name) {
                Ok(pattern) => pattern,
                Err(_) => continue, // Skip unknown patterns
            };

            // Extract the code snippet
            let start_byte = node.start_byte();
            let end_byte = node.end_byte();
            let snippet = source[start_byte..end_byte].to_string();

            // Create span for the block
            let span = Span {
                start: Location::from(node.start_position()),
                end: Location::from(node.end_position()),
            };

            blocks.push(Block::new(
                pattern.into(),
                path_buf.clone(),
                snippet,
                None,
                span,
                Offset { start: start_byte, end: end_byte },
            ));
        }

        Ok(blocks)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_rust_code() {
        // Sample Rust code to parse
        let sample_code = r#"
            struct User {
                name: String,
                age: u32,
            }
            
            enum Status {
                Active,
                Inactive,
                Pending,
            }
            
            impl Status {
                pub fn status(&self) -> &Status {
                    self
                }
            }

            #[test]
            fn process_user(user: &User) -> Status {
                if user.age > 18 {
                    Status::Active
                } else {
                    Status::Pending
                }
            }
        "#;

        let blocks = RustTreeSitter::default()
            .parse("./test.rs", sample_code)
            .unwrap();
        insta::assert_debug_snapshot!(blocks);
    }
}
