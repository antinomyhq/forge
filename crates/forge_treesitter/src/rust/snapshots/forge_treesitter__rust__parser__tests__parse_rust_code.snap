---
source: crates/forge_treesitter/src/rust/parser.rs
expression: blocks
---
[
    Block {
        kind: Struct,
        path: "./test.rs",
        snippet: "struct User {\n                name: String,\n                age: u32,\n            }",
        scope: None,
        span: Span {
            start: Location {
                line: 1,
                column: 12,
            },
            end: Location {
                line: 4,
                column: 13,
            },
        },
        offset: Offset {
            start: 13,
            end: 96,
        },
    },
    Block {
        kind: Enum,
        path: "./test.rs",
        snippet: "enum Status {\n                Active,\n                Inactive,\n                Pending,\n            }",
        scope: None,
        span: Span {
            start: Location {
                line: 6,
                column: 12,
            },
            end: Location {
                line: 10,
                column: 13,
            },
        },
        offset: Offset {
            start: 122,
            end: 224,
        },
    },
    Block {
        kind: Function,
        path: "./test.rs",
        snippet: "pub fn status(&self) -> &Status {\n                    self\n                }",
        scope: None,
        span: Span {
            start: Location {
                line: 13,
                column: 16,
            },
            end: Location {
                line: 15,
                column: 17,
            },
        },
        offset: Offset {
            start: 280,
            end: 356,
        },
    },
    Block {
        kind: Function,
        path: "./test.rs",
        snippet: "fn process_user(user: &User) -> Status {\n                if user.age > 18 {\n                    Status::Active\n                } else {\n                    Status::Pending\n                }\n            }",
        scope: None,
        span: Span {
            start: Location {
                line: 19,
                column: 12,
            },
            end: Location {
                line: 25,
                column: 13,
            },
        },
        offset: Offset {
            start: 404,
            end: 607,
        },
    },
]
